sys/posix/netdb/netdb.c

```diff
@@ -0,0 +1,276 @@
+                | AI_V4MAPPED | AI_ALL | AI_ADDRCONFIG)
+        ) {
+        return EAI_BADFLAGS;
+    }
+
+    if ((hints->ai_flags & AI_CANONNAME) && nodename == NULL) {
+        return EAI_BADFLAGS;
+    }
+
+    if ((hints->ai_family != AF_UNSPEC) &&
+        (hints->ai_family != AF_INET) &&
+        (hints->ai_family != AF_INET6)) {
+        return EAI_FAMILY;
+    }
+
+    if (servname != NULL) { /* service location specified, try to resolve */
+        if (hints->ai_flags & AI_NUMERICSERV) { /* service name specified: convert to port number */
+            port = atoi(servname);
+
+            if (port == 0 && (servname[0] != '0')) {
+                /* atoi failed - service was not numeric */
+                return EAI_SERVICE;
+            }
+
+            if ((port < 0) || (port > (int)0xffff)) {
+                /* invalid port number */
+                return EAI_SERVICE;
+            }
+        }
+        else {
+            /* non-numeric service names not supported */
+            return EAI_SERVICE;
+        }
+    }
+
+    if (nodename != NULL) { /* node location specified, try to resolve */
+        if (hints->ai_flags & AI_NUMERICHOST) { /* no DNS lookup, just parse for an address string */
+
+            if (hints->ai_family == AF_INET) {
+                addr.type = AF_INET;
+            }
+            else if (hints->ai_family == AF_INET6) {
+                addr.type = AF_INET6;
+            }
+            else { /* auto-detect IP version */
+                if (strstr(nodename, ":") != NULL) { /* IPv6 */
+                    addr.type = AF_INET6;
+                }
+                else if (strstr(nodename, ".") != NULL) { /* IPv4 */
+                    addr.type = AF_INET;
+                }
+                else {
+                    /* try IPv6 */
+                    addr.type = AF_INET6;
+                }
+            }
+
+            if (addr.type == AF_INET) {
+                if (inet_pton(AF_INET, nodename, &addr.ipv4) != 1) {
+                    return EAI_NONAME;
+                }
+            }
+            else if (addr.type == AF_INET6) {
+                if (inet_pton(AF_INET6, nodename, &addr.ipv6) != 1) {
+                    return EAI_NONAME;
+                }
+            }
+        }
+        else { /* do actual DNS lookup */
+            uint8_t addr_buf[16] = { 0 };
+            int res = dns_query(nodename, addr_buf, hints->ai_family);
+            switch (res) {
+            case -ETIMEDOUT:
+                /* fall-through */
+            case -EAGAIN:
+                return EAI_AGAIN;
+
+            case -ENOMEM:
+                return EAI_MEMORY;
+
+            case INADDRSZ:
+                addr.type = AF_INET;
+                memcpy(&addr.ipv4, addr_buf, INADDRSZ);
+                break;
+
+            case IN6ADDRSZ:
+                addr.type = AF_INET6;
+                memcpy(&addr.ipv6, addr_buf, IN6ADDRSZ);
+                break;
```

Comment:
Why not use `addr` directly?

Make `ipvx_addr_t` the following:

```C
typedef struct ipvx_addr {
    sa_family_t type;
    union {
        struct in_addr ipv4;
        struct in6_addr ipv6;
        uint8_t u8[16];
    };
} ipvx_addr_t;
```
Then just do
```suggestion
            int res = dns_query(nodename, addr->u8, hints->ai_family);
            switch (res) {
            case -ETIMEDOUT:
                /* fall-through */
            case -EAGAIN:
                return EAI_AGAIN;
            case -ENOMEM:
                return EAI_MEMORY;
            case INADDRSZ:
                addr.type = AF_INET;
                break;
            case IN6ADDRSZ:
                addr.type = AF_INET6;
                break;
```

And safe both a little ROM and RAM